File: tutorial/4.0-violaJones/viola_jones/train.py
from .base import *
from .dataset import *
from .serializer import *
from . import config
import os

def sample_positives(cascade, samples, count):
    positiveSamples = []
    confidences = []

    i = 0; nPicked = 0
    for sample in samples:
        result, conf = predict(cascade, sample)
        if result:
            positiveSamples.append(sample)
            confidences.append(conf)
            nPicked += 1

        i += 1
        if nPicked >= count:
            break

    hitRate = nPicked / i
    return positiveSamples, confidences, hitRate

def append_stage(cascade, positives, negatives, minTPR):
    tpSamples, tpConfs, tprHitRatio = sample_positives(cascade, positives, len(positives))
    print(f"Sampled positives: {len(tpSamples)}/{len(positives)}. Hit rate: {tprHitRatio}")

    nFPsTpPick = 2 * len(positives) - len(tpSamples)
    fpSamples, fpConfs, fprHitRatio = sample_positives(cascade, negatives, nFPsTpPick)
    print(f"Sampled negatives: {len(fpSamples)}/{len(negatives)}. Hit rate: {fprHitRatio}")

    if fprHitRatio <= config.targetFPR:
        return False

    labels = ([+1.0] * len(tpSamples)); labels.extend([-1.0] * len(fpSamples))
    labels = np.array(labels)

    samplesLst = tpSamples; samplesLst.extend(fpSamples)
    samples = np.empty(len(samplesLst), object); samples[:] = samplesLst[:] # retain nested 2D arrays

    confs = tpConfs; confs.extend(fpConfs)
    confs = np.array(confs)

    train_bct(cascade, samples, labels, confs, config.max_trees, minTPR, config.maxFPR)
    return True

def train(cascade_name = "cascade.bin", dataset = "dataset/"):
    print(f"""
Database folder must contain image files and their corresponding labels file (image name with '.txt' ext).
Labels file is in YOLOv3 format. Negative images may omit labels file (if there are not forbidden regions).
Cascade will be written as '{cascade_name}' or read if exists.
          """)
    
    if not os.path.exists(cascade_name):        
        cascade = Cascade(config.wh_ratio, config.tree_depth)
        write_cascade(cascade, cascade_name)

    cascade = read_cascade(cascade_name)
    trainPos = PositiveDataset(dataset, cascade.wh_ratio, RoiRandomJitterTransformer())
    trainNeg = NegativeDataset(dataset, cascade.wh_ratio)

    for stage_idx in range(cascade.stage_count(), len(config.minTPRs)):
        print(f"\n---- Stage {stage_idx + 1} ----\n")
        
        minTPR = config.minTPRs[stage_idx]
        is_appended = append_stage(cascade, trainPos, trainNeg, minTPR)
        if not is_appended:
            break
            
        write_cascade(cascade, cascade_name)       

File: training and runtime app (C++)/src/ViolaJones/Train/Train.cpp
#define PARALLEL 1 //execute training procedure in parallel where applicable

#include "Train.hpp"
#include <Extensions/ConsoleExtensions.h>

using namespace System;
using namespace ViolaJones;

/// @brief Outputs an error produced inside a thread - not caught in a main exception block.
/// @param ex An occurred exception.
static void OutputThreadError(Exception& ex)
{
    Console::Error((string)"Unexpected thread exception.");
    Console::Error(ex);
}

/// @brief Runs the app - parses the arguments and runs a training procedure.
/// @param args Console args.
static void RunApp(List<string>& args)
{
    Thread<>::SubscribeErrorHandler(OutputThreadError);

    if (args.Count() > 1)
        throw Exception((string)"Invalid number of arguments.");

    //db path
    var dbPath = (args.Count() == 1) ? args[0]: DATABASE_PATH;
    
    //config & cascade load
    bool isConfigCreated;
    var config = TrainConfig::LoadOrCreate(dbPath, isConfigCreated);
    if (isConfigCreated) //if created with default values give a user to change values
        return;

    var cascadeFile = Path::Combine(dbPath, CASCADE_FILE_NAME);
    var cascade = Cascade::LoadOrCreate(cascadeFile, config);

    Console::WriteLine((string)"---- Used config:");
    Console::Write(config);
    Console::WriteLine();

    //data init and loading
    Console::WriteLine((string)"Dataset:");
    var transform = RoiRandomJitterTransform();
    var baseSet   = LabeledDataset(dbPath, cascade.WidthHeightRatio);

    var posSet = PositiveDataset(baseSet, &transform);
    var negSet = NegativeDataset(baseSet);

    //training
    for (var i = cascade.StageCount(); i < config.MinTPRs.Count(); i++)
    {
        Console::WriteLine();
        Console::Warning((string)"------- Stage: "  + (i + 1) + " -------");

        var minTPR = config.MinTPRs[i];
        var isStageAppended = TryAppendStage(cascade, posSet, negSet, minTPR, 0.5f, config.MaxFPR, config.MaxTreeCount);
        if (isStageAppended == false)
            break;

        cascade.ToFile(cascadeFile);
    }

    Console::WriteLine((string)"Done!");
}

int main(int argCount, char* argValues[])
{
    Console::ForegroundColor = ConsoleColor::Green;
    Console::WriteLine((string)"Object detection (Viola Jones) training.");

    Console::ForegroundColor = ConsoleColor::Yellow;
    Console::WriteLine((string)"Argument: [database path] = 'database/'");
    Console::WriteLine((string)"\tExample: 'Train database/'");
    Console::WriteLine();

    Console::ForegroundColor = ConsoleColor::Default;
    Console::WriteLine((string)"Database folder must contain image files and their corresponding labels file (image name with '.txt' ext).");
    Console::WriteLine((string)"Labels file is in YOLOv3 format. Negative images will omit labels file.");
    Console::WriteLine((string)"Cascade will be written in '<databasePath>/cascade.bin' file (and read if exists).");
    Console::WriteLine();
    Console::WriteLine();

    try
    {
        var arguments = GetArguments(argCount, argValues);
        RunApp(arguments);
    }
    catch (Exception& ex)
    {
        Console::Error(ex);
        return -1;
    }
    
    return 0;
}